{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "dea1deab",
   "metadata": {},
   "source": [
    "# Phylogenetics in Practice\n",
    "## Preliminaries\n",
    "Make sure to select the R kernel for this iPython notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33e00ac4",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "At the beginning of 2015, several cases of patients presenting symptoms of mild fever, rash, conjunctivitis and joint pain were reported by clinicians in South America. Investigation revealed infections with Zika virus, an obscure relative of dengue and chikungunya, that until the late 2000s had caused sporadic benign infections in Africa and Asia. Outbreaks emerged in Micronesia in 2007 and French Polynesia in 2013-14, during which an unusual increase in the incidence of severe neurological complications was first noticed. Over the course of 2015 and into 2016, clinicians in Brazil reported an upswing in the numbers of children born with microcephaly. The finger of blame was pointed towards Zika virus, a flavivirus, members of which group had been known to exhibit neurotropism and thereby carry the risk of causing neurological defects. As of March 2016, the epidemiological history and expectations for the ongoing spread of Zika virus were very uncertain.\n",
    "\n",
    "The purpose of this practical is to demonstrate the tools available for phylogenetic analysis during early outbreak investigation. The aims are to address the following questions:\n",
    "- What was the geographic origin of the 2015/2016 Zika outbreak?\n",
    "- How was the virus related to previous Zika outbreaks?\n",
    "- When did the 2015/2016 Zika virus strain diverge from other variants?\n",
    "- Was there anything special about the 2015/2016 strain?\n",
    "- At what speed was Zika virus spreading through South America?\n",
    "- Could we have predicted the total number of cases?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9078ef51",
   "metadata": {},
   "source": [
    "## Genome Sequences\n",
    "\n",
    "As of January 2016, only a modest number of Zika virus whole, or near-whole, genome sequences were available:\n",
    "\n",
    "<iframe src=\"https://docs.google.com/spreadsheets/d/e/2PACX-1vS8yll6PC_F0ermshphnLg4ON21mMmGm3FXBOp9IRS3NlGG20L8VusTTryHzrf4PBAvNU1kN_TeK5Mo/pubhtml?widget=true&amp;headers=false\" width=\"100%\"></iframe>\n",
    "\n",
    "Click [here](https://docs.google.com/spreadsheets/d/1tqKzU8rquKbkBbPqmRg6yW2jtt9EuQYjpHKq2mPAgaY/edit?usp=sharing) for the Google Doc. The first step in the analysis is to obtain the whole genome sequences from Genbank using the accession numbers provided in publications or found through a Genbank search query.\n",
    "\n",
    "In your web browser, navigate to https://www.ncbi.nlm.nih.gov/genbank/\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
